### PR TITLE
Forbid os.readlink()

### PR DESCRIPTION
**Pull request by adferrand**
The method `os.readlink()` has a significant behavior change with Python 3.8+ on Windows. 

Starting with this version, it will return the resolved path in its "extended-style" form unconditionally, a form which allows to use more than 259 characters in a Windows path, and its string representation is prepended with "\\\\?\\".

See https://docs.microsoft.com/fr-fr/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#maximum-path-length-limitation

Problem is that `os.readlink()` does it for any path, including paths that could be represented with the normal form. As a consequence, any string comparison with a path provided in the normal form will fail even if it represents the same path. This makes Certbot partially break on Windows with Python 3.8.

My proposition in this PR is to forbid `os.readlink()`, and provide `certbot.compat.filesystem.readlink()` which serves the same purpose at resolving the pointed path of a link, and has a consistent behavior over supported Python versions.

### CODE REVIEW

**Comment by adferrand** [comment]
Sadly we cannot just replace `os.readlink()` with `os.path.realpath()`. The former is designed to get the path that points a link (so basically it resolves links located as the leaf of a path, and this could also be a link) while the later get the real path by recursively resolving any potential link in the path.

Several Certbot features are relying on the specific behavior of `os.readlink()`, so we need to keep it and provide a dedicated `filesystem.readlink()` method.

I updated my PR description.

**Comment by bmw** [comment]
This is needed because tests are failing in https://github.com/certbot/certbot/pull/8465 because the path with `\\?\` and the path without it don't compare equal. We can fix that by stripping `\\?\`, but what happens if the path is long enough that this is needed? Are we going to hit the same problem? If so, is there a more fundamental problem here that needs to be fixed? A couple ideas on what that could look like are changing how we compare paths or erroring out nicely if the path is too long in addition to something like what's in this PR.

What do you think?

**Comment by adferrand** [comment]
In the new `readlink()`function, the path is returned without modifications if it is long enough so that the extended form is required.

So this case is covered.

**Comment by bmw** [comment]
Let me be honest and start by saying that my concern here does feel a bit theoretical and nitpicky so please try and take a step back bear with me, but I do honestly have some concerns about this PR and I feel like we're not addressing the real issue here.

I worry that all software Certbot is built on does not consistently use the extended form once the limit is reached. I'm especially worried about the Python standard library whose support for Windows isn't great as we know.

I suppose a more concrete question I have is what is the path comparison that's failing in https://github.com/certbot/certbot/pull/8465? On Windows, two paths where one is prefexed with `\\?\` do refer to the same path and the code that evaluated them as not equal was incorrect. Properly handling cases like this when comparing paths at least initially seems like the more fundamental problem to me here. I wonder if we shouldn't be fixing that problem there (and elsewhere if it exists) instead of just removing the prefix somewhere else. Something like [pathlib](https://docs.python.org/3/library/pathlib.html) may help with this if the `os` module doesn't.

Whether or not we do this, we may still want to forbid `os.readlink` because of the easy gotcha here, but I think preventing people from using standard library functions is something we should do sparingly and I don't love just reverting to the old behavior of the function. Maybe if we fix the comparison using something like `pathlib`, we'd want to return a `pathlib` object here too to make it easier for people to make the right comparisons in the future.

Alternatively, we can say either that we don't care about fixing the naive comparison(s) that broke this code right now and we'll wait and see if other bugs pop up and/or we don't support Windows paths with 259+ characters, but those options feel like half measures to me which I'd rather avoid especially when I'd be surprised if fixing the comparison was that hard. If we wanted to audit the code for other comparisons, I'd happily save that for future work if you wanted, but just reverting the behavior of the standard library and moving on here honestly doesn't feel like the right call to me.

**Comment by adferrand** [comment]
Well I see your point.

The failing logic is here: https://github.com/certbot/certbot/blob/fcc8b38c02cc10ff320b401ffa242f2253ce9552/certbot/certbot/_internal/account.py#L327

Because `next_dir_path` is generated using other facilities from `os.path` that returns the path in its normal form, the equality comparison fails against the output of `os.readlink`.

I would like to say first that having to handle with long paths on Windows is an extremely corner case, and would certainly not be encountered in real life. First, dealing with long paths on Windows is really hard. Most CLI and GUI operations will fail to create or handle them, you will need to raise admin privileges and go with low-level kernel methods to deal with that. Also, the situations where links are explored are internal situations to Certbot (it is about handling `live` and `archive`). Finally I am almost sure that Python will simply not run from a working directory that requires the long form to be represented: in fact, I tested it and `python --version` cannot even return correctly.

So initially I wanted to create a method in `certbot.compat.filesystem` that would try the path comparison using all combinaisons of extended/normal form. This is surely sufficient and would work to solve the specific problem here.

But from my point of view, this is this comparison solution which seems to be a half-measure. Indeed, similarly to the other forbid we did in `certbot.compat`, we have here an inconsistent behavior of a standard Python method over Python versions and/or OS systems which is the reason to forbid them in the first place. We could expect for instance that other `os.readlink` calls currently existing or added in the future may break silently or not in the Certbot project lifetime.

That's why I took the approach of a new `readlink` method which is consistent other Python versions and OS systems, and forbid the original one. It seems a reasonnable approach, given the number of time `os.readlink` is used (which is low).

Before we go deeper in how to compare paths, what you think of my approach of forbidding `os.readlink` vs handling path comparison inplace?

**Comment by bmw** [comment]
Thanks for the link to the comparison and flagging how little support there is for long paths on Windows including `python --version` not even working if your CWD is a path that is too long which I verified.

I also agree that it makes sense to forbid `os.readlink` to prevent the easy gotcha here in the future. One of my previous ideas was to return something like a `pathlib` object which I believe helps to handle these path comparisons. If Python and other tools support for long paths is so bad though, this is probably overkill.

I still personally don't like just reverting the behavior of the standard library and returning paths prefixed with `\\?\` when they are too long though. If we ever hit this case, I think we'll return the prefixed path and most likely fail later in a way that will be harder to track down. What do you think of one of the ideas from [my first comment on this PR](https://github.com/certbot/certbot/pull/8472#issuecomment-742872724) of in addition to what you have in this PR, erroring out if the path is too long with a message that paths this long is not supported by Certbot?

**Comment by adferrand** [comment]
Yes your proposition about erroring out when long paths are encountered seems fine for me. Anyway given the situation Certbot would be non functional anyway if a long path is involved.

File: certbot/certbot/compat/filesystem.py [hunk 1]
```diff
@@ -383,8 +385,30 @@ def realpath(file_path):
...
+    :raise: ValueError if a long path (260> characters) is encountered on Windows
+    """
+    path = os.readlink(link_path)
+
+    # Max length of a normal path is 260 characters on Windows, including the non printable
+    # termination character "<NUL>". The termination character is not included in Python
+    # strings, giving a max length of 259 characters, + 4 characters for the extended form
+    # prefix, to an effective max length of the studied path of 263 characters.
+    if not POSIX_MODE and path.startswith('\\\\?\\') and len(path) < 264:
+        raise ValueError("Long paths are not supported by Certbot on Windows.")
```
Comment (bmw) [inline certbot/certbot/compat/filesystem.py:405]: I think we only want to error here if the path is long it actually requires the `\\?\` prefix. I'd expect tests on this code to fail on Python 3.8+ on Windows.

I also personally think adding a simple unit test or two for this function is a good idea.

Comment (adferrand) [reply certbot/certbot/compat/filesystem.py:405]: Sorry my implementation is wrong, I could not get the time to continue the work until now.

File: certbot/certbot/compat/filesystem.py [hunk 2]
```diff
@@ -383,8 +385,35 @@ def realpath(file_path):
...
+    :return: The path the symlink points to
+    :returns: str
+    :raise: ValueError if a long path (260> characters) is encountered on Windows
+    """
+    path = os.readlink(link_path)
+
+    if POSIX_MODE or not path.startswith('\\\\?\\'):
+        return path
+
+    # At that point, we are on Windows and the path returned uses the long form (Python 3.8+).
```
Comment (bmw) [inline certbot/certbot/compat/filesystem.py:403]:
```suggestion
    # At this point, we know we are on Windows and that the path returned uses
    # the extended form which is done for all paths in Python 3.8+
```

**Review by bmw — COMMENTED** [review]
One documentation nitpick and assuming the tests I started at https://dev.azure.com/certbot/certbot/_build/results?buildId=3249&view=results to test our behavior with Python 3.8 on Windows passes, this LGTM.

**Review by bmw — APPROVED** [review]

**Comment by bmw** [comment]
Failure in the tests I ran is unrelated to the changes in this PR.

**Pull request merged**